Refactor the partitioning module to make it easier to introduce new algorithms

wesleywiser opened the pull request:
I've split the `librustc_mir::monomorphize::partitioning` module into a few files and introduced a `Partitioner` trait which allows us to decouple the partitioning algorithm from the code which integrates it into the query system. This should allow us to introduce new partitioning algorithms much more easily. I've also gone ahead and added a `-Z` flag to control which algorithm is used (currently there is only the `default`). 

I left a few comments in places where things might be improved further. 

r? @pnkfelix cc @rust-lang/wg-incr-comp

wesleywiser committed:
Split partitioning.rs into a module

wesleywiser committed:
Add partitioning -Z option to control which partitioning scheme is used

bjorn3 commented on src/librustc_mir/monomorphize/partitioning/mod.rs:
```diff
@@ -0,0 +1,433 @@
+
+    match strategy {
+        "default" => Box::new(default::DefaultPartitioning),
+        _ => bug!("unknown partitioning strategy"),
```
`tcx.sess.fatal`?

wesleywiser commented:
Let's make sure this doesn't hurt performance. 

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 42d7cf395b14b1b39b3050ffad364a1e15234f4a with merge 9b74ae087324a980952c4b760b1d7aab32b3ce65...
<!-- homu: {"type":"TryBuildStarted","head_sha":"42d7cf395b14b1b39b3050ffad364a1e15234f4a","merge_sha":"9b74ae087324a980952c4b760b1d7aab32b3ce65"} -->

wesleywiser commented on src/librustc_mir/monomorphize/partitioning/mod.rs:
```diff
@@ -0,0 +1,433 @@
+//! Partitioning Codegen Units for Incremental Compilation
```
This comment block is taken verbatim from its original location. Parts of it describe the partitioning process over all and parts are specific to the algorithm (now called `DefaultPartitioner`). 

This could probably be left as is for now but eventually we should split it into two parts and move the bits about the algorithm into module documentation for the `DefaultPartitioner` algorithm.

wesleywiser commented on src/librustc_mir/monomorphize/partitioning/mod.rs:
```diff
@@ -0,0 +1,433 @@
+        target_cgu_count: usize,
+    );
+
+    fn place_inlined_mono_items(
```
This name & signature was taken from the existing code. One that that was not immediately obvious to me was that this does a lot more than just add copies of `#[inline]` functions into CGUs. This function is primary responsible for instantiating monomorphized functions in the modules that call them. (Perhaps that's what was meant by "inlined"?)

wesleywiser commented on src/librustc_session/options.rs:
```diff
@@ -939,6 +939,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "support compiling tests with panic=abort (default: no)"),
     parse_only: bool = (false, parse_bool, [UNTRACKED],
         "parse only; do not compile, assemble, or link (default: no)"),
+    partitioning_strategy: Option<String> = (None, parse_opt_string, [TRACKED],
```
Maybe this should be `-Zcgu-partitioning-strategy` instead?

davidtwco replied:
I think `cgu-partitioning-strategy` would be best.

wesleywiser replied:
I agree (and went with `cgu-partitioning-strategy`)

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/863831063), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=34485)
Build commit: 9b74ae087324a980952c4b760b1d7aab32b3ce65 (`9b74ae087324a980952c4b760b1d7aab32b3ce65`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=34485","checks-actions":"https://github.com/rust-lang-ci/rust/runs/863831063"},"merge_sha":"9b74ae087324a980952c4b760b1d7aab32b3ce65"} -->

rust-timer commented:
Queued 9b74ae087324a980952c4b760b1d7aab32b3ce65 with parent 9d09331e00b02f81c714b0c41ce3a38380dd36a2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9d09331e00b02f81c714b0c41ce3a38380dd36a2&end=9b74ae087324a980952c4b760b1d7aab32b3ce65).

rust-timer commented:
Finished benchmarking try commit (9b74ae087324a980952c4b760b1d7aab32b3ce65): [comparison url](https://perf.rust-lang.org/compare.html?start=9d09331e00b02f81c714b0c41ce3a38380dd36a2&end=9b74ae087324a980952c4b760b1d7aab32b3ce65).

bjorn3 commented:
<0.1% speedup overall.

wesleywiser commented:
r? @pnkfelix cc @rust-lang/wg-incr-comp

davidtwco commented on src/librustc_mir/monomorphize/partitioning/merging.rs:
```diff
@@ -0,0 +1,110 @@
+
+use crate::monomorphize::partitioning::PreInliningPartitioning;
+
+pub fn merge_codegen_units<'tcx>(
```
I think this makes sense - I don't think it would prohibit us from having partitioning strategies which don't have a merging phase.

davidtwco reviewed (approved):
Current review comments notwithstanding, LGTM.

wesleywiser commented:
Responded to review feedback so far.

bors commented:
:umbrella: The latest upstream changes (presumably #74214) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74091) made this pull request unmergeable. Please resolve the merge conflicts.

richkadel commented on src/librustc_mir/monomorphize/partitioning/mod.rs:
```diff
@@ -0,0 +1,433 @@
+
+    // Next we try to make as many symbols "internal" as possible, so LLVM has
+    // more freedom to optimize.
+    if !tcx.sess.opts.cg.link_dead_code {
```
if tcx.sess.opts.cg.link_dead_code != Some(true) {

richkadel commented on src/librustc_mir/monomorphize/partitioning/mod.rs:
```diff
@@ -0,0 +1,433 @@
+            }
+        }
+        None => {
+            if tcx.sess.opts.cg.link_dead_code {
```
tcx.sess.opts.cg.link_dead_code == Some(true)

richkadel reviewed (commented):
Sorry but I had to change this option's structure.

wesleywiser commented:
Rebased.

@pnkfelix if you are interested in doing the review, I don't mind waiting for that, otherwise we can find someone else from wg-incr-comp to do the review.

pnkfelix commented:
I left feedback over here on [zulip](https://zulip-archive.rust-lang.org/241847tcompilerwgincrcomp/67460RefactorthepartitioningmodulePR74275.html); basically, I asked wesley to refactor the commit series so that the code shuffling is its own commit, so that the diff will better show what the actual changes are regarding the new trait.

richkadel commented:
No problem. I'll do that.

richkadel commented:
Oops. Wrong PR. Please ignore.

wesleywiser committed:
Add a CGU partitioning trait

This will allow us to prototype different partitioning schemes without
adding a lot of extra conditionals everywhere.

wesleywiser commented:
@pnkfelix I rebased and split out the trait into the first commit. The second commit has no functional changes, just moving the code into a new module.

pnkfelix commented on src/librustc_mir/monomorphize/partitioning.rs:
```diff
@@ -452,248 +413,334 @@ fn default_visibility(tcx: TyCtxt<'_>, id: DefId, is_generic: bool) -> Visibilit
+            internalization_candidates,
+        } = initial_partitioning;
+
+        let single_codegen_unit = initial_cgus.len() == 1;
```
Viewing in diff mode with whitespace changes hidden is a huge win here. :)

pnkfelix commented:
@bors r+

bors commented:
:pushpin: Commit 98b943e7d821aff17fefbf1a094503f9ae172a4a has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 98b943e7d821aff17fefbf1a094503f9ae172a4a -->
<!-- homu: {"type":"Approved","sha":"98b943e7d821aff17fefbf1a094503f9ae172a4a","approver":"pnkfelix"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 1000, this pull request will be tested once the tree is reopened

pietroalbini commented:
@bors treeclosed-

bors commented:
:hourglass: Testing commit 98b943e7d821aff17fefbf1a094503f9ae172a4a with merge 8ba22504e8e5dcbbe136d97f63c1280dabc523d0...
<!-- homu: {"type":"BuildStarted","head_sha":"98b943e7d821aff17fefbf1a094503f9ae172a4a","merge_sha":"8ba22504e8e5dcbbe136d97f63c1280dabc523d0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1028380167), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36218)
Approved by: pnkfelix
Pushing 8ba22504e8e5dcbbe136d97f63c1280dabc523d0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pnkfelix","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36218","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1028380167"},"merge_sha":"8ba22504e8e5dcbbe136d97f63c1280dabc523d0"} -->
